pkg/build/controller/build/build_controller.go

```diff
@@ -388,34 +409,65 @@ func (bc *BuildController) resolveOutputDockerImageReference(build *buildapi.Bui
 			var ok bool
 			streamName, tag, ok = imageapi.SplitImageStreamTag(streamName)
 			if !ok {
-				return "", fmt.Errorf("the referenced image stream tag is invalid: %s", outputTo.Name)
+				return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced image stream tag is invalid: %s", outputTo.Name)
 			}
 			tag = ":" + tag
 		}
 		stream, err := bc.imageStreamStore.ImageStreams(namespace).Get(streamName)
 		if err != nil {
 			if errors.IsNotFound(err) {
-				return "", fmt.Errorf("the referenced output image stream %s/%s does not exist", namespace, streamName)
+				return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced output image stream %s/%s does not exist", namespace, streamName)
 			}
-			return "", fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
+			return "", build.Spec.Output.PushSecret, fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
 		}
 		if len(stream.Status.DockerImageRepository) == 0 {
 			e := fmt.Errorf("the image stream %s/%s cannot be used as the output for build %s/%s because the integrated Docker registry is not configured and no external registry was defined", namespace, outputTo.Name, build.Namespace, build.Name)
 			bc.recorder.Eventf(build, kapi.EventTypeWarning, "invalidOutput", "Error starting build: %v", e)
-			return "", e
+			return "", build.Spec.Output.PushSecret, e
 		}
 		ref = fmt.Sprintf("%s%s", stream.Status.DockerImageRepository, tag)
 	}
-	return ref, nil
+
+	pushSecret = build.Spec.Output.PushSecret
+	if pushSecret == nil {
+		serviceAccount := build.Spec.ServiceAccount
+		if len(serviceAccount) == 0 {
+			serviceAccount = bootstrappolicy.BuilderServiceAccountName
+		}
+		sa, err := bc.kubeClient.Core().ServiceAccounts(build.Namespace).Get(serviceAccount, metav1.GetOptions{})
+		if err != nil {
+			return ref, build.Spec.Output.PushSecret, fmt.Errorf("Error getting push/pull secrets for service account %s/%s: %v", build.Namespace, serviceAccount, err)
```

Comment:
We're ignoring failures where we couldn't get one of the secrets on the SA, we have no idea if that secret is even relevant to image pushing, so it's not necessarily fatal for us not to be able to get it and we can't assume it eventually will exist (or matters for the build at all), so it doesn't make sense to hold the build until all the secrets associated w/ the SA can be retrieved.

Not being able to get the SA itself is immediately fatal(well, will leave the build in a new state and get retried) because someone said "run with this SA+its secrets" and we can't do so.

Anyway no it's not intended to handle the race condition you describe.